File: benchmarks/lz77/java/Lz77.java
import java.util.ArrayList;
import java.util.List;

public class Lz77 {
    static final int WINDOW = 4096;
    static final int MAX_MATCH = 255;
    static final int MIN_MATCH = 3;
    static final int HASH_SIZE = 8192;

    static class Token {
        boolean isMatch;
        int offset;
        int length;
        byte literal;
    }

    static int hash3(byte[] data, int i) {
        return (((data[i] & 0xFF) * 131 + (data[i + 1] & 0xFF)) * 131 + (data[i + 2] & 0xFF)) & (HASH_SIZE - 1);
    }

    static List<Token> encode(byte[] data) {
        int n = data.length;
        int[] hashTable = new int[HASH_SIZE];
        java.util.Arrays.fill(hashTable, -1);

        List<Token> tokens = new ArrayList<>();
        int i = 0;
        while (i < n) {
            int bestLen = 0;
            int bestCand = -1;
            if (i + 3 <= n) {
                int h = hash3(data, i);
                int cand = hashTable[h];
                if (cand != -1 && i - cand <= WINDOW) {
                    int matchLen = 0;
                    while (matchLen < MAX_MATCH && i + matchLen < n && data[cand + matchLen] == data[i + matchLen]) {
                        matchLen++;
                    }
                    if (matchLen >= MIN_MATCH) {
                        bestLen = matchLen;
                        bestCand = cand;
                    }
                }
                hashTable[h] = i;
            }
            Token t = new Token();
            if (bestLen >= MIN_MATCH) {
                t.isMatch = true;
                t.offset = i - bestCand;
                t.length = bestLen;
                tokens.add(t);
                i += bestLen;
            } else {
                t.isMatch = false;
                t.literal = data[i];
                tokens.add(t);
                i += 1;
            }
        }
        return tokens;
    }

    static byte[] decode(List<Token> tokens, int expectedLen) {
        byte[] out = new byte[expectedLen];
        int len = 0;
        for (Token t : tokens) {
            if (t.isMatch) {
                int start = len - t.offset;
                for (int k = 0; k < t.length; k++) {
                    out[len] = out[start + k];
                    len++;
                }
            } else {
                out[len] = t.literal;
                len++;
            }
        }
        return out;
    }

    public static void main(String[] args) {
        int n = args.length > 0 ? Integer.parseInt(args[0]) : 5_000_000;

        byte[] pattern = new byte[64];
        for (int i = 0; i < 64; i++) pattern[i] = (byte) ((i * 7 + 3) % 251);
        byte[] data = new byte[n];
        for (int i = 0; i < n; i++) {
            byte v = pattern[i % 64];
            if (i % 97 == 0) v = (byte) (((v & 0xFF) + 1) % 256);
            data[i] = v;
        }

        List<Token> tokens = encode(data);
        byte[] decoded = decode(tokens, n);

        if (decoded.length != n) {
            System.err.println("self-check failed: roundtrip length mismatch");
            System.exit(1);
        }
        for (int i = 0; i < n; i++) {
            if (decoded[i] != data[i]) {
                System.err.println("self-check failed: roundtrip byte mismatch at " + i);
                System.exit(1);
            }
        }

        long compressedBytes = 0;
        for (Token t : tokens) {
            compressedBytes += t.isMatch ? 4 : 2;
        }

        System.out.println(compressedBytes);
    }
}

File: benchmarks/lz77/rust/lz77.rs
use std::env;

const WINDOW: i64 = 4096;
const MAX_MATCH: usize = 255;
const MIN_MATCH: usize = 3;
const HASH_SIZE: usize = 8192;

struct Token {
    is_match: bool,
    offset: i64,
    length: usize,
    literal: u8,
}

fn hash3(data: &[u8], i: usize) -> usize {
    (((data[i] as usize) * 131 + data[i + 1] as usize) * 131 + data[i + 2] as usize) & (HASH_SIZE - 1)
}

fn lz77_encode(data: &[u8]) -> Vec<Token> {
    let n = data.len();
    let mut hash_table = vec![-1i64; HASH_SIZE];
    let mut tokens = Vec::new();
    let mut i: usize = 0;
    while i < n {
        let mut best_len = 0usize;
        let mut best_cand: i64 = -1;
        if i + 3 <= n {
            let h = hash3(data, i);
            let cand = hash_table[h];
            if cand != -1 && (i as i64 - cand) <= WINDOW {
                let cand_u = cand as usize;
                let mut match_len = 0usize;
                while match_len < MAX_MATCH && i + match_len < n && data[cand_u + match_len] == data[i + match_len] {
                    match_len += 1;
                }
                if match_len >= MIN_MATCH {
                    best_len = match_len;
                    best_cand = cand;
                }
            }
            hash_table[h] = i as i64;
        }
        if best_len >= MIN_MATCH {
            tokens.push(Token { is_match: true, offset: i as i64 - best_cand, length: best_len, literal: 0 });
            i += best_len;
        } else {
            tokens.push(Token { is_match: false, offset: 0, length: 0, literal: data[i] });
            i += 1;
        }
    }
    tokens
}

fn lz77_decode(tokens: &[Token]) -> Vec<u8> {
    let mut out: Vec<u8> = Vec::new();
    for t in tokens {
        if t.is_match {
            let start = out.len() as i64 - t.offset;
            for k in 0..t.length {
                let b = out[(start + k as i64) as usize];
                out.push(b);
            }
        } else {
            out.push(t.literal);
        }
    }
    out
}

fn main() {
    let n: usize = env::args().nth(1).map(|s| s.parse().unwrap()).unwrap_or(5_000_000);

    let mut pattern = [0u8; 64];
    for i in 0..64 {
        pattern[i] = ((i * 7 + 3) % 251) as u8;
    }
    let mut data = vec![0u8; n];
    for i in 0..n {
        let mut v = pattern[i % 64];
        if i % 97 == 0 {
            v = v.wrapping_add(1);
        }
        data[i] = v;
    }

    let tokens = lz77_encode(&data);
    let decoded = lz77_decode(&tokens);

    assert_eq!(decoded.len(), n, "self-check failed: roundtrip length mismatch");
    assert_eq!(decoded, data, "self-check failed: roundtrip byte mismatch");

    let compressed_bytes: i64 = tokens.iter().map(|t| if t.is_match { 4 } else { 2 }).sum();
    println!("{}", compressed_bytes);
}
